worlds/marioland2/logic.py

```diff
@@ -0,0 +1,608 @@
+    return has_pipe_right(state, player)
+
+
+def turtle_zone_3_coins(state, player, coins):
+    return state.has_any(["Water Physics", "Mushroom", "Fire Flower", "Carrot"], player) or coins <= 51
+
+
+def hippo_zone_normal_or_secret_exit(state, player):
+    return (state.has_any(["Hippo Bubble", "Water Physics"], player)
+            or (state.has("Carrot", player)
+                and not is_auto_scroll(state, player, "Hippo Zone")))
+
+
+def hippo_zone_coins(state, player, coins):
+    auto_scroll = is_auto_scroll(state, player, "Hippo Zone")
+    # This is all somewhat forgiving.
+    reachable_coins = 4
+    if auto_scroll:
+        if state.has("Hippo Bubble", player):
+            reachable_coins = 160
+        elif state.has("Carrot", player):
+            reachable_coins = 90
+        elif state.has("Water Physics", player):
+            reachable_coins = 28
+    else:
+        if state.has_any(["Water Physics", "Hippo Bubble", "Carrot"], player):
+            reachable_coins += 108
+            if state.has_any(["Mushroom", "Fire Flower", "Hippo Bubble"], player):
+                reachable_coins += 6
+        if state.has_all(["Fire Flower", "Water Physics"], player):
+            reachable_coins += 1
+        if state.has("Hippo Bubble", player):
+            reachable_coins += 52
+    return coins <= reachable_coins
+
+
+def space_zone_1_normal_exit(state, player):
+    # It is possible, however tricky, to beat the Moon Stage without Carrot or Space Physics.
+    # However, it requires somewhat precisely jumping off enemies. Enemy shuffle may make this impossible.
+    # Instead, I will just always make one or the other required, since it is difficult without them anyway.
+    return state.has_any(["Space Physics", "Carrot"], player)
+
+
+def space_zone_1_secret_exit(state, player):
+    # One or the other is actually necessary for the secret exit.
+    return state.has_any(["Space Physics", "Carrot"], player) and not is_auto_scroll(state, player, "Space Zone 1")
+
+
+def space_zone_1_coins(state, player, coins):
+    auto_scroll = is_auto_scroll(state, player, "Space Zone 1")
+    if auto_scroll:
+        reachable_coins = 12
+        if state.has_any(["Carrot", "Space Physics"], player):
+            reachable_coins += 20
+        # If you have Space Physics, you can't make it up to the upper section. We have to assume you might have it,
+        # so the coins up there must be out of logic if there is auto scrolling.
+        if state.has("Space Physics", player):
+            reachable_coins += 40
+        return coins <= reachable_coins
+    return (coins <= 21 or (coins <= 50 and state.has_any(["Mushroom", "Fire Flower"], player))
+            or state.has_any(["Carrot", "Space Physics"], player))
+
+
+def space_zone_2_midway_bell(state, player):
+    return state.has_any(["Space Physics", "Space Zone 2 Midway Bell", "Mushroom", "Fire Flower",
+                          "Carrot"], player),
```

Comment:
This can just be one line
```suggestion
    return state.has_any(["Space Physics", "Space Zone 2 Midway Bell", "Mushroom", "Fire Flower", "Carrot"], player)
```